pkg/securitycontext/admission/admission.go

```diff
@@ -0,0 +1,177 @@
+package admission
+
+import (
+	kadmission "github.com/GoogleCloudPlatform/kubernetes/pkg/admission"
+	kapi "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/errors"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/auth/user"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/securitycontextconstraints"
+)
+
+func init() {
+	kadmission.RegisterPlugin("SecurityContextConstraint", func(client client.Interface, config io.Reader) (kadmission.Interface, error) {
+		return NewConstraint(client), nil
+	})
+}
+
+type constraint struct {
+	*kadmission.Handler
+	client client.Interface
+}
+
+var _ kadmission.Interface = &constaint{}
+
+func NewConstraint(client client.Interface) kadmission.Interface {
+	return &constraint{
+		Handler: kadmission.NewHandler(kadmission.Create, kadmission.Update),
+		client:  client,
+	}
+}
+
+func (c *constraint) Admit(a kadmission.Attributes) error {
+	// 1.  find all SCCs the pod has access to:
+	// 2.  Fully resolve each SCC
+	//     1.  determine uid from the SCC's run as user policy
+	//     2.  set the uid on the SCC
+	// 3.  Match the pod's SC to an SCC; for each SCC:
+	//     1.  generate the SC
+	//     2.  validate against the pod's SC
+
+	if a.GetResource() != "pods" {
+		return nil
+	}
+
+	pod, ok := a.GetObject().(kapi.Pod)
+	if !ok {
+		return errors.NewBadRequest("a pod was received, but could not convert the request object.")
+	}
+
+	ns := a.GetNamespace()
+	matchedConstraints, err := getMatchingSecurityContextConstraints(c.client, ns, pod)
+	if err != nil {
+		return err
+	}
+
+	for _, constraint := range matchedConstraints {
+		if constraint.RunAsUser.Type == kapi.RunAsUserStrategyMustRunAs && constraint.RunAsUser.UID == nil {
+			uid, err := getPreallocatedUID(constraint)
+			if err != nil {
+				return err
+			}
+
+			constraint.RunAsUser.UID = &uid
+		}
+
+		if constraint.SELinuxContext.Type == kapi.SELinuxStrategyMustRunAs && constraint.SELinuxContext.SELinuxOptions.Level == "" {
+			level, err := getPreallocatedLevel(constraint)
+			if err != nil {
+				return err
+			}
+
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+	}
```

Comment:
Yeah, we can punt on it for now though.  I'm focusing on the broad strokes for now.